### PR TITLE
Use `safe_infer` in `_unpack_args` and `_unpack_keywords`

### PR DESCRIPTION
**Pull request by tusharsadhwani**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Fixes https://github.com/pylint-dev/pylint/issues/8544

### CODE REVIEW

**Comment by tusharsadhwani** [comment]
This is causing one test for me to fail locally, but seems not relevant:

```
=================================== FAILURES ===================================
_______________ AstroidManagerTest.test_module_is_not_namespace ________________

self = <tests.test_manager.AstroidManagerTest testMethod=test_module_is_not_namespace>

    def test_module_is_not_namespace(self) -> None:
        self.assertFalse(util.is_namespace("tests.testdata.python3.data.all"))
        self.assertFalse(util.is_namespace("__main__"))
>       self.assertFalse(
            util.is_namespace(list(EXT_LIB_DIRS)[0].rsplit("/", maxsplit=1)[-1]),
        )
E       AssertionError: True is not false

tests/test_manager.py:138: AssertionError
=========================== short test summary info ============================
FAILED tests/test_manager.py::AstroidManagerTest::test_module_is_not_namespace - AssertionError: True is not false
====== 1 failed, 1483 passed, 50 skipped, 14 xfailed, 1 xpassed in 12.88s ======
```

**Review by Pierre-Sassoulas — DISMISSED** [review]
Thanks ! I feel like there's a lot of place that could use safe_infer too in astroid.

**Review by DanielNoord — APPROVED** [review]

File: astroid/arguments.py [hunk 1]
```diff
@@ -134,14 +135,7 @@ def _unpack_args(self, args, context: InferenceContext | None = None):
...
             if isinstance(arg, nodes.Starred):
-                try:
-                    inferred = next(arg.value.infer(context=context))
-                except InferenceError:
-                    values.append(Uninferable)
-                    continue
-                except StopIteration:
-                    continue
-
+                inferred = safe_infer(arg.value, context=context)
```
Comment (jacobtylerwalls) [inline astroid/arguments.py:138]: Isn't there the same flaw for unpacking `**kwargs` in `_unpack_keywords` just above this?

Comment (jacobtylerwalls) [reply astroid/arguments.py:138]: I know we have to draw the line somewhere, but it's essentially the same test case :-)

Comment (tusharsadhwani) [reply astroid/arguments.py:138]: Alright, I'll go ahead and fix that too :D

**Comment by tusharsadhwani** [comment]
A test failed due to the change it seems:

```
        node = extract_node(code)
        assert isinstance(node, nodes.NodeNG)
        result = node.inferred()
        assert len(result) == 2
        assert isinstance(result[0], nodes.Dict)
>       assert result[1] is util.Uninferable
E       assert <Dict.dict l.4 at 0x113aa9610> is Uninferable
E        +  where Uninferable = util.Uninferable

tests/test_inference.py:1458: AssertionError
____________________________________ AstroidManagerTest.test_module_is_not_namespace ____________________________________
```

File: tests/test_inference.py [hunk 2]
```diff
@@ -5296,6 +5305,36 @@ def test_duplicated_keyword_arguments(self) -> None:
...
+                if x == '1':
+                    nums = (1, 2)
+                return nums
+
+            def add(x, y):
+                return x + y
+
+            nums = get_nums()
+
+            kwargs = {foo: bar, 1: baz}
```
Comment (jacobtylerwalls) [inline tests/test_inference.py:5321]: I think you need to introduce some ambiguity by assigning `kwargs = {}` before reassigning it to have content, like you did with `nums`.

File: tests/test_inference.py [hunk 3]
```diff
@@ -5296,6 +5305,36 @@ def test_duplicated_keyword_arguments(self) -> None:
...
+            if nums:
+                add(*nums)
+                print(**kwargs)
+        """
+        # Test that `*nums` argument should be Uninferable
+        ast = parse(code, __name__)
+        *_, add_call, print_call = list(ast.nodes_of_class(nodes.Call))
+        nums_arg = add_call.args[0]
+        add_call_site = self._call_site_from_call(add_call)
+        assert add_call_site._unpack_args([nums_arg]) == [Uninferable]
```
Comment (jacobtylerwalls) [inline tests/test_inference.py:5332]: I suggest matching the surrounding style with using `self.assertEqual`

**Comment by jacobtylerwalls** [comment]
> A test failed due to the change it seems:

The new result shows an improved inference result. The test was added just to test a crash, and it's being unnecessarily specific in comparing against `Uninferable`. I'll push an update.

**Comment by jacobtylerwalls** [comment]
This has the potential to raise more pylint messages as inference is changing, so I don't think it should be backported.

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Could you add a changelog entry for 2.16?

File: ChangeLog [hunk 4]
```diff
@@ -52,6 +52,8 @@ Release date: TBA
 
 * Remove dependency on ``wrapt``.
 
+* ``CallSite._unpack_args`` and ``CallSite._unpack_keywords`` now use ``safe_infer()`` for
+  better inference and fewer false positives.
```
Comment (jacobtylerwalls) [inline ChangeLog:56]:
```suggestion
  better inference and fewer false positives.

  Closes pylint-dev/pylint#8544
```

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**